lib/vnet/process_manager.go

```diff
@@ -0,0 +1,85 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+func newProcessManager() (*ProcessManager, context.Context) {
+	ctx, cancel := context.WithCancel(context.Background())
+	g, ctx := errgroup.WithContext(ctx)
+	pm := &ProcessManager{
+		g:      g,
+		cancel: cancel,
+		closed: make(chan struct{}),
+	}
+	pm.closeOnce = sync.OnceFunc(func() {
+		close(pm.closed)
+	})
+	return pm, ctx
+}
+
+// ProcessManager handles background tasks needed to run VNet.
+// Its semantics are similar to an error group with a context, but it cancels the context whenever
+// any task returns prematurely, that is, a task exits while the context was not canceled.
+type ProcessManager struct {
+	g         *errgroup.Group
+	cancel    context.CancelFunc
+	closed    chan struct{}
+	closeOnce func()
+}
+
+// AddCriticalBackgroundTask adds a function to the error group. [task] is expected to block until
+// the context returned by [newProcessManager] gets canceled. The context gets canceled either by
+// calling Close on [ProcessManager] or if any task returns.
+func (pm *ProcessManager) AddCriticalBackgroundTask(name string, task func() error) {
+	pm.g.Go(func() error {
+		err := task()
+		if err == nil {
+			// Make sure to always return an error so that the errgroup context is canceled.
+			err = fmt.Errorf("critical task %q exited prematurely", name)
+		}
+		return trace.Wrap(err)
+	})
+}
+
+// Wait blocks and waits for the background tasks to finish, which typically happens when another
+// goroutine calls Close on the process manager.
+func (pm *ProcessManager) Wait() error {
+	err := pm.g.Wait()
+	select {
+	case <-pm.closed:
+		// Errors are expected after the process manager has been closed,
+		// usually due to context cancellation.
+		return nil
```

Comment:
What problems does this solve compared to the version on master? I assume it has to do with what errors are returned by `Wait`, but by just looking at this file alone is hard to tell what changed in the whole process orchestration compared to master.

Do we just want `Close()` to not make `Wait()` return a "context cancelled" error? I know we've changed how we approach this specific error a couple of times, so I'd just want to have some clarity on how to approach it in the future. ;)

https://github.com/gravitational/teleport/blob/47bf4cad41f106f40b7dee25883d345b0a85e340/lib/vnet/run.go#L177-L218

Reply:
i don't feel strongly about this, for some reason while writing i wanted it not to return an error after the processmanager was closed, if you have another opinion i can revert

Reply:
No opinion, just wanted to understand the reasoning.